### PR TITLE
user authenticated: Closes #23

### LINKED ISSUE #23: 16: Authenticated user security
Background: An authenticated user

As an Authenticated User 
I cannot view another user's private data (current or past orders, etc) 
I cannot view the administrator screens or use admin functionality 
I cannot make myself an admin

### PR DESCRIPTION
**Pull request by swdonovan**
…k a before_action and use either current user or a similar method would be wise.

### CODE REVIEW

**Review by jimszalew — APPROVED** [review]
I like your idea of using a current_user method to refactor, but this looks good.

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -2,5 +2,11 @@ class UsersController < ApplicationController
 
   def show
     @user = User.find(params[:id])
+    if session[:user_id] == @user.id
+      render 'show'
+    else
+      session[:user_id] = nil
+      redirect_to login_path
+    end
```
Comment (jimszalew) [inline app/controllers/users_controller.rb:10]: I really like this setting the user id to nil.

Comment (MikelSage) [reply app/controllers/users_controller.rb:10]: Agreed. It's reminiscent of what Facebook did to Mike when he tried changing stuff in his cookies.

**Review by MikelSage — COMMENTED** [review]
I think this is pretty great. In keeping with the user story though I think it'd be nice if you built out the skeleton of the admin functionality. Just the name spaced controllers and a skeleton dashboard and build out a current_admin check.

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (dphilla) [reply app/controllers/users_controller.rb:10]: I like this too. The more comfortable I get with sessions, the more I realize that this is the way to go.

**Review by dphilla — COMMENTED** [review]
Looks good to me!

**Comment by swdonovan** [comment]
Hey can someone take a look at this when they get time and let me know what they think about this.  I put the login and out in the drop down menu and put a Admin Dashboard link for admins as well in the drop down ONLY if they are an admin and doesn't show if they aren't an admin.  I left it like this for the purpose that if there was a bug and a user that wasn't an admin could access admin functionality, they would have less of a chance of knowing about it due to the dashboard being in the drop down.  Better solution would be to put this functionality in the footer of the page.

File: app/views/layouts/_navbar.html.erb [hunk 2]
```diff
@@ -25,20 +25,19 @@
...
+            <% if current_user %>
+              <% if current_admin %>
+                <li><a href="<%=  admin_user_path(current_user.id)  %>">Admin Dashboard</a></li>
+              <% end %>
+            <li><a href="#"> <%= "Logged in as #{current_user.username}" %> </a></li>
+            <li> <%= link_to "Logout", login_path, method: :delete %></li>
+          <% else %>
+            <li><a href="/users/new">Signup</a></li>
+            <li><a href="/login">Login</a></li>
+          <% end  %>
```
Comment (swdonovan) [inline app/views/layouts/_navbar.html.erb:40]: Here is the drop down that is in there.  I can move this back around to the normal if needed

File: app/controllers/admin/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      redirect_to login_path
+    end
+  end
+
+  # def update
+  #   if session_check?(@user)
+  #     render 'edit'
+  #   else
+  #     session[:user_id] = nil
+  #     flash[:danger] = "GET OUT"
```
Comment (MikelSage) [inline app/controllers/admin/users_controller.rb:19]: That was a great movie.

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (swdonovan) [reply app/controllers/users_controller.rb:10]: Sessions and JS will be fun to play with next mod and mod 4 :)

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,6 +1,6 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
-  helper_method :set_cart, :categories, :current_user
+  helper_method :set_cart, :categories, :current_user, :current_admin
```
Comment (swdonovan) [inline app/controllers/application_controller.rb:3]: does the session check need to be helper method or not? couldn't decide?

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -13,4 +13,12 @@ def categories
   def current_user
     @user ||= User.find(session[:user_id]) if session[:user_id]
   end
+
+  def current_admin
+    current_user && current_user.admin?
+  end
+
+  def session_check?(user)
```
Comment (MikelSage) [inline app/controllers/application_controller.rb:21]: Interesting use of a session[:user_id] here.

**Review by MikelSage — APPROVED** [review]
Looking good and ready to merge.

**Pull request merged**